desc/compute/_field.py

```diff
@@ -4,800 +4,1858 @@
-            + data["B_theta"] * data["B^theta_t"]
-            + data["B^zeta"] * data["B_zeta_t"]
-            + data["B_zeta"] * data["B^zeta_t"]
-        )
-    if has_dependencies("grad(|B|^2)_zeta", params, transforms, profiles, data):
-        data["grad(|B|^2)_zeta"] = (
-            data["B^theta"] * data["B_theta_z"]
-            + data["B_theta"] * data["B^theta_z"]
-            + data["B^zeta"] * data["B_zeta_z"]
-            + data["B_zeta"] * data["B^zeta_z"]
-        )
-
-    # gradient vector
-    if has_dependencies("grad(|B|^2)", params, transforms, profiles, data):
-        data["grad(|B|^2)"] = (
-            data["grad(|B|^2)_rho"] * data["e^rho"].T
-            + data["grad(|B|^2)_theta"] * data["e^theta"].T
-            + data["grad(|B|^2)_zeta"] * data["e^zeta"].T
-        ).T
-
-    # magnitude
-    if has_dependencies("|grad(|B|^2)|/2mu0", params, transforms, profiles, data):
-        data["|grad(|B|^2)|/2mu0"] = (
-            jnp.sqrt(
-                data["grad(|B|^2)_rho"] ** 2 * data["g^rr"]
-                + data["grad(|B|^2)_theta"] ** 2 * data["g^tt"]
-                + data["grad(|B|^2)_zeta"] ** 2 * data["g^zz"]
-                + 2 * data["grad(|B|^2)_rho"] * data["grad(|B|^2)_theta"] * data["g^rt"]
-                + 2 * data["grad(|B|^2)_rho"] * data["grad(|B|^2)_zeta"] * data["g^rz"]
-                + 2
-                * data["grad(|B|^2)_theta"]
-                * data["grad(|B|^2)_zeta"]
-                * data["g^tz"]
-            )
-            / 2
-            / mu_0
         )
+        + (data["B^theta_z"] * data["B^zeta"] + data["B^theta"] * data["B^zeta_z"])
+        * (
+            dot(data["e_theta_t"], data["e_zeta"])
+            + dot(data["e_zeta_t"], data["e_theta"])
+        )
+        + data["B^theta"]
+        * data["B^zeta"]
+        * (
+            dot(data["e_theta_tz"], data["e_zeta"])
+            + dot(data["e_zeta_tz"], data["e_theta"])
+            + dot(data["e_theta_t"], data["e_zeta_z"])
+            + dot(data["e_zeta_t"], data["e_theta_z"])
+        )
+    ) / data["|B|"] - data["|B|_t"] * data["|B|_z"] / data["|B|"]
+    return data
+
 
+@register_compute_fun(
+    name="<|B|>_vol",
+    label="\\langle |B| \\rangle_{vol}",
+    units="T",
+    units_long="Tesla",
+    description="Volume average magnetic field",
+    dim=0,
+    params=[],
+    transforms={"grid": []},
+    profiles=[],
+    function_of="",
+    data=["sqrt(g)", "|B|", "V"],
+)
+def _B_vol(params, transforms, profiles, data, **kwargs):
+    data["<|B|>_vol"] = (
+        jnp.sum(transforms["grid"].weights * data["sqrt(g)"] * data["|B|"]) / data["V"]
+    )
     return data
 
 
-def compute_magnetic_tension(params, transforms, profiles, data=None, **kwargs):
-    """Compute magnetic tension."""
-    data = compute_contravariant_current_density(
-        params,
-        transforms,
-        profiles,
-        data=data,
-        **kwargs,
+@register_compute_fun(
+    name="<|B|>",
```

Comment:
The flux surface average definitions I have read in papers (and D'haeseleer's flux coords book) are usually equivalent to
$$\langle Q \rangle\ = \dfrac{\iint d\theta\ d\zeta\ \sqrt{g}\ Q\}{\iint d\theta\ d\zeta\  \sqrt{g}\}$$

The denominator is the derivative of signed volume wrt rho. The jacobean `sqrt(g)` is not enforced to be positive.

For the magnetic well, we take volume to be considered always positive. With this convention the flux surface average formula becomes
$$\langle Q \rangle\ = \dfrac{\iint d\theta\ d\zeta\ \lvert \sqrt{g} \rvert\ Q\}{\iint d\theta\ d\zeta\  \lvert \sqrt{g} \rvert\}$$

1. Since we are moving  `<B^2>` and `<B^2>_r` from the magnetic well compute function into the `data_index`, maybe we should add something like 'uses magnitude of jacobean' to the description?
2. `<|B|>` was not used in the magnetic well compute function. Is `<|B|>` planned to be used in a context with the magnitude of jacobean in the average, or do we want that for consistency?

Reply:
would it matter, since the sign of the jacobian would cancel anyways?

Reply:
If the jacobean sign is constant throughout the whole vector you're right that it won't matter. Otherwise they could give different results

Reply:
Well if the surfaces are nested it should have a constant sign throughout the volume, so it shouldn't be an issue. And once #202 is merged it should be fine also.